File: python/탐욕법(그리디)/구명보트.py

""" Level 2. 구명보트 """ 

def solution(people, limit):
    answer = 0

    people.sort()

    start, target = 0, len(people) - 1
    while start <= target:

        p1 = people[start]
        while start <= target and p1 + people[target] > limit:
            answer += 1
            target -= 1

        if start <= target:
            answer += 1

        start += 1
        target -= 1

    return answer

print(solution([70, 50, 80, 50], 100))      # 3
print(solution([70, 80, 50], 100))          # 3

print(solution([70], 100))                  # 1

File: Java/src/그리디/구명보트.java
package 그리디;

import java.util.*;

public class 구명보트 {

    public static void main(String[] args) {

        int[] p1 = {70, 50, 80, 50};
        int l1 = 100;
        System.out.println(solution(p1, l1));

        int[] p2 = {70, 80, 50};
        int l2 = 100;
        System.out.println(solution(p2, l2));

        int[] p3 = {50};
        int l3 = 100;
        System.out.println(solution(p3, l3));
    }

    public static int solution(int[] people, int limit) {
        
        Arrays.sort(people);
        
        int start = 0;
        int end = people.length - 1;
        
        int answer = 0;
        while (start <= end) {
            
            while (start <= end && people[start] + people[end] > limit) {
                answer++;
                end--;
            }
            
            if (start <= end) {
                answer++;
            }

            start++;
            end--;
        }

        return answer;
    }
}